flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java

```diff
@@ -540,14 +540,26 @@ protected int getCurrentOuterSnapshotVersion() {
 
         @Override
         protected void readOuterSnapshot(
-                int readOuterSnapshotVersion, DataInputView in, ClassLoader userCodeClassLoader) {
-            readVersion = readOuterSnapshotVersion;
+                int readOuterSnapshotVersion, DataInputView in, ClassLoader userCodeClassLoader)
+                throws IOException {
+            if (readOuterSnapshotVersion <= LAST_VERSION_WITHOUT_ROW_KIND) {
+                supportsRowKind = false;
+            } else if (readOuterSnapshotVersion == LAST_VERSION_WITHOUT_ROW_KIND + 1) {
```

Comment:
I think adding another new flag instead of `LAST_VERSION_WITHOUT_ROW_KIND + 1` would be better to understand.